src/include/duckdb/function/udf_function.hpp

```diff
@@ -0,0 +1,303 @@
+//===----------------------------------------------------------------------===//
+//                         DuckDB
+//
+// duckdb/function/udf_function2.hpp
+//
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include "duckdb/common/types.hpp"
+#include "duckdb/function/scalar_function.hpp"
+#include "duckdb/main/client_context.hpp"
+#include "duckdb/parser/parsed_data/create_scalar_function_info.hpp"
+#include "duckdb/common/types/data_chunk.hpp"
+#include "duckdb/transaction/transaction.hpp"
+#include <string>
+#include <vector>
+#include <type_traits>
+
+using namespace std;
+
+namespace duckdb {
+
+struct UDFWrapper {
+public:
+	template<typename TR, typename... Args>
+	static scalar_function_t CreateScalarFunction(string name, TR (*udf_func)(Args...)) {
+		const std::size_t num_template_argc = sizeof...(Args);
+		switch(num_template_argc) {
+			case 1:
+				return CreateUnaryFunction<TR, Args...>(name, udf_func);
+			case 2:
+				return CreateBinaryFunction<TR, Args...>(name, udf_func);
+			case 3:
+				return CreateTernaryFunction<TR, Args...>(name, udf_func);
+			default:
+				throw duckdb::NotImplementedException("UDF function only supported until ternary!");
+		}
+	}
+
+	template<typename TR, typename... Args>
+	static scalar_function_t CreateScalarFunction(string name, vector<SQLType> args, SQLType ret_type, TR (*udf_func)(Args...)) {
+		if(!TypesMatch<TR>(ret_type)) {
+			throw duckdb::TypeMismatchException(GetTypeId<TR>(), GetInternalType(ret_type),
+					"Return type doesn't match with the first template type.");
+		}
+
+		const std::size_t num_template_types = sizeof...(Args);
+		if(num_template_types != args.size()) {
+			throw duckdb::InvalidInputException("The number of templated types should be the same quantity of the SQLType arguments.");
+		}
+
+		switch(num_template_types) {
+			case 1:
+				return CreateUnaryFunction<TR, Args...>(name, args, ret_type, udf_func);
+			case 2:
+				return CreateBinaryFunction<TR, Args...>(name, args, ret_type, udf_func);
+			case 3:
+				return CreateTernaryFunction<TR, Args...>(name, args, ret_type, udf_func);
+			default:
+				throw duckdb::NotImplementedException("UDF function only supported until ternary!");
+		}
+	}
+
+	template<typename TR, typename... Args>
+	static void RegisterFunction(string name, scalar_function_t udf_function, ClientContext &context) {
+	    vector<SQLType> arguments;
+	    GetArgumentTypesRecursive<Args...>(arguments);
+
+	    SQLType ret_type = GetArgumentType<TR>();
+
+	    RegisterFunction(name, arguments, ret_type, udf_function, context);
+	}
+
+	static void RegisterFunction(string name, vector<SQLType> args, SQLType ret_type,
```

Comment:
Could you move this function to a source file (`udf_function.cpp`)? This should allow you to remove the includes for `duckdb/parser/parsed_data/create_scalar_function_info.hpp` and `duckdb/main/client_context.hpp`.